File: examples/python/structured_task.py
"""Local structured task with one explicit repair; no Provider, Run or Root binding."""
import asyncio
import json
from dataclasses import replace
from purra.api import AgentModelTask, AgentModelTaskRunner, IntegrationCheck, StructuredOutputContract, check_integration
from purra.contracts import AgentMessage, ModelCompletion, ModelRequest, ModelTokenUsage
from purra.model_invocation import AgentModelInvocationManager, ModelInvocationContext
from purra.model_protocol import generic_capability_snapshot

class FixtureGateway:
    calls = 0
    async def complete(self, messages, invocation, signal=None):
        self.calls += 1
        return ModelCompletion(AgentMessage('assistant', 'invalid' if self.calls == 1 else '{"ok":true}'),
            model='fixture', finish_reason='stop', usage=ModelTokenUsage(10, 5),
            applied_generation_limit=invocation.max_generation_tokens)
    async def stream(self, *args): raise AssertionError('structured completion expected')

async def main():
    gateway = FixtureGateway()
    model = ModelRequest('fixture', 'fixture', replace(generic_capability_snapshot(), max_generation_tokens=128))
    runner = AgentModelTaskRunner(AgentModelInvocationManager(gateway), ModelInvocationContext('example'), model)
    output = StructuredOutputContract('example', '1', {'type':'object', 'properties':{'ok':{'type':'boolean'}}, 'required':['ok'], 'additionalProperties':False})
    async def probe():
        result = await runner.complete_structured((AgentMessage('user', 'Return an ok flag.'),), AgentModelTask(model), output=output, repair_attempts=1)
        assert result.value == {'ok':True} and result.receipt.attempts == 2
        assert result.receipt.persistence == 'none' and result.receipt.root_budget == 'not_bound'
        assert gateway.calls == 2
    report = await check_integration(component='structured-example', version='1',
        checks=[IntegrationCheck('structured_output', 'deterministic', probe)])
    assert next(row for row in report['checks'] if row['capability']=='structured_output' and row['category']=='deterministic')['status'] == 'passed'
    print(json.dumps(report, indent=2))

if __name__ == '__main__': asyncio.run(main())

File: typescript/examples/structured-task.ts
// Local structured task with one explicit repair; no Provider, Run or Root binding.
import { ModelTaskRunner, StructuredOutputContract, checkIntegration, type ModelCapabilitySnapshot, type ModelGateway } from "purra";
const capabilities: ModelCapabilitySnapshot = {
  schemaVersion: 2,
  profileId: "example:structured-task",
  providerProtocol: "custom",
  contextWindowTokens: 16_000,
  maxGenerationTokens: 512,
  thinkingTokenAccounting: "unknown",
  protocol: {
    reasoningControl: "selectable", reasoningReplay: "ignored", toolCalling: "supported",
    requiredToolChoice: "supported", parallelToolCalls: "supported", streaming: "unavailable",
    cancellation: "supported", assistantContentWithToolCalls: "optional", jsonSchemaLevel: "unknown",
    streamFinishSemantics: "normalized", usageSemantics: "normalized",
  },
};
let calls=0;
const model: ModelGateway={capabilities,async invoke(request){
  calls++;
  return {message:{role:"assistant",content:calls===1?'invalid':'{"ok":true}'},finishReason:"stop",
    appliedGenerationLimit:request.outputBudget.maxGenerationTokens,
    usage:{inputTokens:10,generationTokens:5,totalTokens:15}};
}};
const runner=new ModelTaskRunner({runId:"example",model});
const output=await StructuredOutputContract.create({schemaId:"example",schemaVersion:"1",
  schema:{type:"object",properties:{ok:{type:"boolean"}},required:["ok"],additionalProperties:false}});
const report=await checkIntegration({component:"structured-example",version:"1",checks:[{
  capability:"structured_output",category:"deterministic",async probe(){
    const result=await runner.completeStructured([{role:"user",content:"Return an ok flag."}],{output,repairAttempts:1});
    if (JSON.stringify(result.value)!=='{"ok":true}' || result.receipt.attempts!==2 || calls!==2) throw Error("fixture failed");
    if (result.receipt.persistence!=="none" || result.receipt.rootBudget!=="not_bound") throw Error("missing authority was inferred");
  }
}]});
if(report.checks.find(row=>row.capability==="structured_output"&&row.category==="deterministic")?.status!=="passed") throw Error("fixture failed");
console.log(JSON.stringify(report,null,2));
